app/scripts/modules/core/src/managed/Environments.tsx

```diff
@@ -1,5 +1,59 @@
-import React from 'react';
+import React, { useState, useEffect } from 'react';
 
-export default function Environments() {
-  return <div>For there shall be no greater pursuit than that towards desired state.</div>;
+import { Application, ApplicationDataSource } from '../application';
+import { IManagedApplicationSummary } from '../domain/IManagedEntity';
+import { ColumnHeader } from '../presentation/layout/ColumnHeader';
+import { ArtifactsList } from './ArtifactsList';
+import { EnvironmentsList } from './EnvironmentsList';
+
+import styles from './Environments.module.css';
+import { isEqual } from 'lodash';
+
+const CONTENT_WIDTH = 1200;
+
+export interface ISelectedArtifact {
+  name: string;
+  version: string;
+}
+
+interface IEnvironmentsProps {
+  app: Application;
+}
+
+export function Environments(props: IEnvironmentsProps) {
+  const { app } = props;
+  const dataSource: ApplicationDataSource<IManagedApplicationSummary<
+    'resources' | 'artifacts' | 'environments'
+  >> = app.getDataSource('environments');
+  const [selectedArtifact, setSelectedArtifact] = useState<ISelectedArtifact>();
```

Comment:
I'd prefer to use the data source's observable than manage refreshes manually.


I have a `useObservable` hook in a pending branch I'm working on.  What do you think of this:

`useObservable.hook.ts`:
```ts
/** Subscribes the callback to the observable on mount. Unsubscribes on unmount. */
function useObservable<T>(observable: Observable<T>, callback: (val: T) => void): void {
  useEffect(() => {
    const subscription = observable.subscribe(callback);
    return () => subscription.unsubscribe();
  }, [observable]);
}

function useObservableValue<T>(observable: Observable<T>): void {
  const [value, setValue] = useState<T>(undefined);
  useObservable(observable, setValue);
  return value;
}
```

Then in this file:
```ts
const environments = useObservableValue(dataSource.data$);
```

Reply:
I like it. I was thinking the same thing and had a brief @erikmunson about creating a `useDataSource` hook, but we didn't think it would end up being used in too many places.

The thinking was that we wouldn't want a ton of components grabbing data directly from the datasources and instead would prefer for them to be managed by a root component and fed through the component tree.

Thoughts?

Reply:
yeah this is really similar to what we were brainstorming on, and if we ever wanted one that was specific to data sources we could just build it on top of this hook. Let's wait until the hook makes it into a PR + gets merged and then switch over to it?

Reply:
I think a really thin `useDataSource` could possibly be good if it provides more functionality than the generic observable one, such as the ability to request a refresh.  Otherwise, if it's just streaming values, I think `useObservableValue` is clear enough.  I'd love it if all usage of datasource transitioned towards using the `data$` and away from the `onRefresh` api.

> managed by a root component and fed through the component tree

agreed

Reply:
> Let's wait until the hook makes it into a PR + gets merged

I'd just write the hooks now.  That PR I referenced is _way back burner_ and may never even land.